src/web/js/beforePyret.js

```diff
@@ -48,15 +48,127 @@ window.stickMessage = function(message) {
   var err = $("<div>").addClass("active").text(message);
   $(".notificationArea").prepend(err);
 };
+window.mkWarningUpper = function(){return $("<div class='warning-upper'>");}
+window.mkWarningLower = function(){return $("<div class='warning-lower'>");}
 
 $(window).bind("beforeunload", function() {
   return "Because this page can load slowly, and you may have outstanding changes, we ask that you confirm before leaving the editor in case closing was an accident.";
 });
+
+CodeMirror.defineDocExtension("rangeHandle", function(position, options) {
+  var inclusiveLeft,
+      inclusiveRight,
+      shared;
+  if (options === undefined) {
+    inclusiveLeft = false,
+    inclusiveRight = false;
+    shared = true;
+  } else {
+    if (options.inclusiveLeft !== undefined)
+      inclusiveLeft = options.inclusiveLeft;
+    if (options.inclusiveRight !== undefined)
+      inclusiveRight = options.inclusiveRight;
+    if (options.shared !== undefined)
+      shared = options.shared;
+  }
+  return this.markText(position.from, to,
+    { inclusiveLeft   : inclusiveLeft,
+      inclusiveRight  : inclusiveRight,
+      shared          : shared  });
+});
+
+var Position = function() {
+  function Position(doc, from, to) {
+    this.doc  = doc;
+    this.changeGeneration = doc.changeGeneration();
+    this.from = from;
+    this.to   = to;
+  }
+}();
+
+var Documents = function() {
```

Comment:
I don't particularly like this pattern of `var Thing` and then an IIFE with a `function Thing` inside it.  The shadowing is pretty odd to think about, IMO.